---
description: 'Disallow using the spread operator when it might cause unexpected behavior.'
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> 🛑 This file is source code, not the primary documentation location! 🛑
>
> See **https://typescript-eslint.io/rules/no-misused-spread** for documentation.

The [spread operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax) (`...`) is a powerful feature in JavaScript that can be misused in ways not always detectable by TypeScript. This rule disallows using the spread operator on types where spreading can lead to unexpected behavior.

## Examples

<Tabs>
<TabItem value="❌ Incorrect">

```ts
declare const userName: string;
const chars = [...userName];

declare const arr: number[];
const arrSpread = { ...arr };

declare const set: Set<number>;
const setSpread = { ...set };

declare const map: Map<string, number>;
const mapSpread = { ...map };

declare function getObj(): { a: 1; b: 2 };
const getObjSpread = { ...getObj };
```

</TabItem>
<TabItem value="✅ Correct">

```ts
declare const userName: string;
const chars = userName.split('');

declare const arr: number[];
const arrSpread = [...arr];

declare const set: Set<number>;
const setSpread = [...set];

declare const map: Map<string, number>;
const mapObject = Object.fromEntries(map);

declare function getObj(): { a: 1; b: 2 };
const getObjSpread = { ...getObj() };
```

</TabItem>
</Tabs>

## Options

### `allowStrings`

By default, this rule disallows using the spread operator on strings inside arrays.
Set `"allowStrings": true` to allow this behavior.

<Tabs>
<TabItem value="❌ Incorrect">

```ts
const fruit = 'Apple';
const chars = [...fruit];
```

</TabItem>

<TabItem value="✅ Correct">

```ts option='{"allowStrings": true}'
// { "allowStrings": true }
const fruit = 'Apple';
const chars = [...fruit];
```

</TabItem>
</Tabs>
```

### `allowFunctions`

By default, this rule disallows using the spread operator on functions without additional properties.
Set `"allowFunctions": true` to allow this behavior.

<Tabs>
<TabItem value="❌ Incorrect">

```ts
declare function doWork();

const spread = { ...doWork };
```

</TabItem>

<TabItem value="✅ Correct">

```ts option='{"allowFunctions": true}'
// { "allowFunctions": true }
declare function doWork();

const spread = { ...doWork };
```

</TabItem>
</Tabs>

### `allowClassInstances`

By default, this rule disallows using the spread operator on instances of classes.
Set `"allowClassInstances": true` to allow this behavior.

<Tabs>
<TabItem value="❌ Incorrect">

```ts
class User {
  name: string;

  constructor(name: string) {
    this.name = name;
  }
}

const user = new User('John');

const spread = { ...user };
```

</TabItem>

<TabItem value="✅ Correct">

```ts option='{"allowClassInstances": true}'
// { "allowClassInstances": true }
class User {
  name: string;

  constructor(name: string) {
    this.name = name;
  }
}

const user = new User('John');

const spread = { ...user };
```

</TabItem>
</Tabs>

### `allowClassDeclarations`

By default, this rule disallows using the spread operator on class declarations.
Set `"allowClassDeclarations": true` to allow this behavior.

<Tabs>
<TabItem value="❌ Incorrect">

```ts
class User {
  name: string;

  constructor(name: string) {
    this.name = name;
  }
}

const spread = { ...User };
```

</TabItem>

<TabItem value="✅ Correct">

```ts option='{"allowClassDeclarations": true}'
// { "allowClassDeclarations": true }
class User {
  name: string;

  constructor(name: string) {
    this.name = name;
  }
}

const spread = { ...User };
```

</TabItem>
</Tabs>

### `allowForKnownSafeIterables`

By default, this rule disallows using the spread operator on iterables. This
option allows marking specific iterable types as "safe" to be spread.

This option takes an array of type specifiers to consider safe.
Each item in the array must have one of the following forms:

- A type defined in the current file, or globally available (`"SafeIterable"`)
- A type defined in a file (`{ from: "file", name: "SafeIterable", path: "src/safe-iterable.ts" }` with `path` being an optional path relative to the project root directory)
- A type from the default library (`{ from: "lib", name: "CustomIterable" }`)
- A type from a package (`{ from: "package", name: "AwesomeIterable", package: "iterable-lib" }`, this also works for types defined in a typings package).

Examples of a configuration for this option:

```json
"@typescript-eslint/no-misused-spread": [
  "error",
  {
    "allowForKnownSafeIterables": [
      "SafeIterable",
      { "from": "file", "name": "ProbablySafeIterable" },
      { "from": "lib", "name": "CustomIterable" },
      { "from": "package", "name": "AwesomeIterable", "package": "iterable-lib" }
    ]
  }
]
```

<Tabs>
<TabItem value="❌ Incorrect">

```ts
type UnsafeIterable = Iterable<number>;

declare const iterable: UnsafeIterable;

const spread = { ...iterable };
```

</TabItem>
<TabItem value="✅ Correct">

```ts option='{"allowForKnownSafeIterables":["SafeIterable"]}'
type SafeIterable = Iterable<number>;

declare const iterable: SafeIterable;

const spread = { ...iterable };
```

</TabItem>
</Tabs>

## When Not To Use It

If you intentionally want to use the spread operator in those cases, and expect
the specific behavior that comes with it, you might not want this rule.
For example, when you want to spread an array into an object and expect the
result to be an object with the array elements as values and the array indices
as keys.
